RP2350: Add RISCV, Unify RP2350, Update UART, Add XH3IRQ Interrupt Controller

AnnsAnns opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md.
-->

### Contribution description

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->

After two failed attempts, I finally managed to get a shared and unified setup working for the RP2350. As opposed to the other approaches I took, where I tried to fully unify both the RISCV and ARM versions, this version takes a different approach similar to cortexm_common or riscv_common where each CPU still exists as a separate entity, e.g. `rp2350_arm` and `rp2350_riscv` but uses the shared `rp2350_common` folder.

This PR supersedes both #21745 and #21746, as in it:

* adds support for the Hazard3 CPU used on the RP2350
* adds support for the XH3IRQ Interrupt Controller used by the Hazard3. While fairly similar to CLIC/PLIC, the Hazard3 on the RP2350 actually has a custom interrupt controller to "mimick" the way interrupts work on the Cortex ARM side, even having support for direct vector table jumps (Not enabled here since I wanted to still use the trap_entry/trap_handler of riscv_common.
* updates the uart
* enables interrupts on the CortexM33
* unifies both the RISCV and ARM side of the RP2350
* renames rp2350 (arm version) to rp2350_arm
* renames rpi-pico-2 (arm version) to rpi-pico-2-arm
* fixes debugging completely (GDB now runs smoothly both on ARM and RISCV)

However, as of now, it does not yet add the multicore example. All the points raised in the other PRs also still hold true.

### Testing procedure

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->

Remember to specify `USEMODULE += stdio_uart` if you wish to print via UART.

For my testing I mostly modified `examples/basic/blinky` to also print. On RISCV you can also use `xh3irq_force_irq` to force interrupts for testing purposes, for example, on each blinky loop, if you feel like it :)

#### RISCV
`make PROGRAMMER=openocd QUIET=0 BOARD=rpi-pico-2-riscv flash` flashes the pico 2 in RISCV mode using openocd, you can also use `make PROGRAMMER=picotool QUIET=0 BOARD=rpi-pico-2-riscv flash` to flash it using picotool.

#### ARM
`make PROGRAMMER=openocd QUIET=0 BOARD=rpi-pico-2-arm flash` flashes the pico 2 in RISCV mode using openocd, you can also use `make PROGRAMMER=picotool QUIET=0 BOARD=rpi-pico-2-arm flash` to flash it using picotool.

### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

AnnsAnns committed:
boards/rpi-pico-2-riscv: add riscv based rp2350 board

AnnsAnns committed:
cpu/rp2350_riscv: add initial support

AnnsAnns committed:
cpu/riscv_common: Add periph xh3irq support

AnnsAnns committed:
cpu/rp2350: move to new uart

AnnsAnns committed:
cpu/rp2350_common: create shared folder for RISCV & ARM

AnnsAnns committed:
boards/rpi-pico-2: adjust to new rp2350_arm name

crasbe commented:
Would you mind creating a "DELETEME" commit that adds the two boards to the `.murdock` file, so that the CI tests actually are meaningful? :)

crasbe commented on cpu/rp2350_common/periph/uart.c:
```diff
@@ -0,0 +1,243 @@
+
+void isr_uart1(void) {
+    isr_handler(1);
+    #ifdef RP2350_USE_ARM
+    cortexm_isr_end();
+    #endif
```
```suggestion
#ifdef RP2350_USE_ARM
    cortexm_isr_end();
#endif
```

Considering that the code is shared, it would make sense to have this as a runtime check when using RISC-V and ARM at the same time (maybe. eventually.)

crasbe commented on cpu/rp2350_common/vectors.c:
```diff
@@ -76,9 +87,14 @@ WEAK_DEFAULT void isr_spareirq_3(void);
+ * @see 3.2 Interrupts and IRQn_Type in RP2350.h
+ */
+#ifdef RP2350_USE_ARM
+/* CortexM does some macro magic, RISCV doesnt */
```
```suggestion
/* CortexM does some macro magic, RISCV does not */
```

crasbe commented on cpu/rp2350_common/vectors.c:
```diff
@@ -132,5 +148,3 @@ const isr_t vector_cpu[CPU_IRQ_NUMOF] = {
     (void*)isr_spareirq_5,   /* 51 SPAREIRQ_IRQ_5 */
 };
-
-/** @} */
```
This should probably not be removed?

crasbe replied:
No?

crasbe commented on cpu/rp2350_riscv/Makefile:
```diff
@@ -0,0 +1,7 @@
+MODULE = cpu
+
+# add a list of subdirectories, that should also be built
+DIRS = $(RIOTCPU)/rp2350_common $(RIOTCPU)/riscv_common
```
```suggestion
DIRS += $(RIOTCPU)/rp2350_common
DIRS += $(RIOTCPU)/riscv_common
```

crasbe commented on cpu/rp2350_riscv/Makefile.features:
```diff
@@ -0,0 +1,9 @@
+CPU_CORE := rv32imac
```
```suggestion
CPU_CORE    := rv32imac
```

crasbe commented on cpu/rp2350_riscv/Makefile.features:
```diff
@@ -0,0 +1,9 @@
+
+FEATURES_PROVIDED += periph_xh3irq
+
+include $(RIOTCPU)/rp2350_common/Makefile.features
+
+include $(RIOTCPU)/riscv_common/Makefile.features
```
```suggestion
include $(RIOTCPU)/rp2350_common/Makefile.features
include $(RIOTCPU)/riscv_common/Makefile.features
```

crasbe commented on cpu/rp2350_riscv/picobin_block.s:
```diff
@@ -0,0 +1,37 @@
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+/* Picobin block required for the binary */
+/* This defines the minimum viable image metadata to be recognized by the RP2350 bootloader */
+/* based on RP2350 Chapter 5.9.1 */
```
```suggestion
/* Picobin block required for the binary.
 * This defines the minimum viable image metadata to be recognized by the RP2350 bootloader
 * based on RP2350 Chapter 5.9.1
 */
```

Same for the other comment block and also the other file.

crasbe commented on features.yaml:
```diff
@@ -283,8 +283,10 @@ groups:
-        - name: cpu_rp2350
-          help: The MCU is a Raspberry Pi RP2350
+        - name: cpu_rp2350_arm
+          help: The MCU is a Raspberry Pi RP2350 CortexM33 ARMM
```
```suggestion
          help: The MCU is a Raspberry Pi RP2350 Cortex-M33 ARM,
```

crasbe commented on features.yaml:
```diff
@@ -283,8 +283,10 @@ groups:
+        - name: cpu_rp2350_arm
+          help: The MCU is a Raspberry Pi RP2350 CortexM33 ARMM
+        - name: cpu_rp2350_riscv
+          help: The MCU is a Raspberry Pi RP2350 Hazard3 RISCV
```
```suggestion
          help: The MCU is a Raspberry Pi RP2350 Hazard3 RISC-V.
```

mguetschow commented on cpu/rp2350_common/Makefile.dep:
```diff
@@ -0,0 +1,4 @@
+USEMODULE += periph
+USEPKG += picosdk
+# We tell the build system that common needs to be built
+USEMODULE += rp2350_common
```
Seems strange to me that the cpu `rp2350_common` needs the module `rp2350_common`. Did you take that from other CPUs? Maybe its also just another weirdness of the build system.

crasbe replied:
> Maybe its also just another weirdness of the build system.

It is a weirdness of the build system. Remember my work when I fixed the Adafruit nRF52 Bootloader Common module stuff? That had exactly the same thing going on...

mguetschow reviewed (dismissed):
Great work! :tada: Some comments from briefly skimming over the changes.

AnnsAnns commented on boards/rpi-pico-2-riscv/board.c:
```diff
@@ -0,0 +1,15 @@
+#include "board.h"
+
+void board_init(void) {
+    /* Re-enable the LED0 pin
+     * Otherwise the LED will not work after a reset
+     * This is needed, esp. when the LED is used via
+     * the define macros */
+    gpio_init(LED0_PIN_ID, GPIO_OUT);
```
The RP2350 introduced a isolation bit, you have to specifically allow output on the pin before you can turn it off and on, I'm not sure how other boards handle it but for this specific instance simply initializing the LED beforehand felt like the simplest solution

AnnsAnns commented on boards/rpi-pico-2-arm/include/periph_conf.h:
```diff
@@ -0,0 +1,51 @@
+#endif
+
+/**
+ * @brief   Configuration details for an UART interface needed by the RPX0XX peripheral
```
There were some slight differences (often in the names of defines, etc. sind mine often came directly from RP2350.h) but through macro magic it should technically be possible yeah. Aside from time my biggest problem with that would be that uh ... I dont have a Pico 1 here anymore :sweat_smile:

AnnsAnns commented on boards/rpi-pico-2-riscv/doc.md:
````diff
@@ -0,0 +1,116 @@
+
+Please refer to the
+[Debugprobe documentation](https://www.raspberrypi.com/documentation/microcontrollers/debug-probe.html#getting-started)
+for more details.
+
+Note that Raspberry Pi actually uses their own OpenOCD fork, which is available
+in the [RP2040 OpenOCD repository](https://github.com/raspberrypi/openocd).
+While technically you can use the standard OpenOCD,
+it is recommended to use the Raspberry Pi fork for better compatibility with the
+RP2350, as its still fairly "new" and under development,
+which is why even their own Pico SDK Extension
+uses the Raspberry Pi fork of OpenOCD, instead of the standard one.
+
+To do this, you need to connect the board to your computer
+and use the following command:
+
+```bash
+PROGRAMMER=openocd BOARD=rpi-pico-2-riscv make flash
+```
+
+You can then debug your application using GDB with the following command:
+
+```bash
+PROGRAMMER=openocd BOARD=rpi-pico-2-riscv make debug
+```
+
+### Flashing using Picotool
+
+Simply connect the board to your computer via USB and use the following command:
+
+```bash
+BOARD=rpi-pico-2-riscv make flash
+```
+
+This is the default method for flashing the Raspberry Pi Pico 2.
+However, it does not allow for debugging using GDB.
+
+@note When programming the board with the Picotool for the first time,
+RIOT will download and install the Picotool locally in the RIOT folder.
+This process will take some minutes to complete.
````
Will do

AnnsAnns commented on cpu/riscv_common/irq_arch.c:
```diff
@@ -32,10 +32,14 @@
 #include "architecture.h"
 
+#if MODULE_PERIPH_XH3IRQ || DOXYGEN
+#include "xh3irq.h"
```
I don't really understand this styling, the hashtag is supposed to stay on one line with the defines? I know that C doesn't really care about whitespaces but this feels wrong to my head :sweat_smile:

AnnsAnns commented on cpu/riscv_common/periph/xh3irq.c:
```diff
@@ -0,0 +1,57 @@
+
+uint32_t xh3irq_has_pending(void) {
+    /* Get MEIP at 0x344 which is the external interrupt pending bit */
+    uint32_t meip = (read_csr(0x344) >> 11) & 0x1;
+
+    return (meip != 0);
```
Bittest functions?

crasbe replied:
https://doc.riot-os.org/bit_8h.html#ac5ef33d3209014d0e19862f91a4ada46

AnnsAnns replied:
I think Ive used it correctly? Not too sure :sweat_smile: https://github.com/RIOT-OS/RIOT/pull/21753/commits/0179c95009c3c48d646d967a400368945953acd2

AnnsAnns commented on cpu/rp2350_common/periph/uart.c:
```diff
@@ -0,0 +1,243 @@
+    switch (parity) {
+    case UART_PARITY_NONE:
+        break;
+    // case UART_PARITY_EVEN:
+    //     io_reg_atomic_set(&dev->UARTLCR_H, UART0_UARTLCR_H_EPS_Msk | UART0_UARTLCR_H_PEN_Msk);
+    //     break;
+    // case UART_PARITY_ODD:
+    //     io_reg_atomic_set(&dev->UARTLCR_H, UART0_UARTLCR_H_PEN_Msk);
+    //     break;
```
Uh quite frankly because I never tested these configurations and there were still issues with different define names between cpu/rp2040 and RP2350.h

AnnsAnns reviewed (commented):
Thank you for all the reviews (esp. so quickly :racehorse:), I will try to go through all the points mentioned and fix them :frog: 👉👉

crasbe commented on cpu/rp2350_arm/Makefile:
```diff
@@ -0,0 +1,7 @@
+MODULE = cpu
+
+# add a list of subdirectories, that should also be built
+DIRS = $(RIOTCPU)/rp2350_common $(RIOTCPU)/cortexm_common
```
This is not changed yet? Did something go wrong with the fixups?

crasbe commented on boards/rpi-pico-2-arm/include/board.h:
```diff
@@ -0,0 +1,42 @@
+ * @author          Tom Hert <git@annsann.eu>
+ */
+
+ #include "cpu.h"
+ #include "periph_conf.h"
+
+ #include "periph/gpio.h"
```
```suggestion
#include "cpu.h"
#include "periph_conf.h"

#include "periph/gpio.h"
```

Visual Studio doing Visual Studio things?

AnnsAnns replied:
This time this can be blamed on Zed I guess (actually probably clangd)

crasbe commented on cpu/riscv_common/irq_arch.c:
```diff
@@ -83,20 +87,45 @@ void riscv_irq_init(void)
+
+    if ((mcause & ~MCAUSE_INT) <= 0xb) {
+        const char *error_messages[] = {
+            "Instruction alignment: Does not occur on RP2350, because 16-bit compressed instructions are implemented, and it is impossible to jump to a byte-aligned address.",
```
```suggestion
            "Instruction alignment: Does not occur on RP2350, because "
            "16-bit compressed instructions are implemented, and it is "
            "impossible to jump to a byte-aligned address.",
```

Perhaps you can split the other lines up too? C allows you to do that:
https://stackoverflow.com/a/797351

AnnsAnns replied:
Updated the entire function to be a bit less hacky (The error messages were straight out of the rp2350 documentation, the new one follows the official RISC-V specs https://riscv.github.io/riscv-isa-manual/snapshot/privileged/#mcause)

AnnsAnns committed:
boards/rpi-pico-2: unify doc.md into one singular doc.md

AnnsAnns commented on boards/rpi-pico-2-riscv/doc.md:
```diff
@@ -0,0 +1,116 @@
+
+More peripherals will be added in the future.
+It should also be noted that we currently only support the Cortex M33 cores,
+not the RISC-V Hazard cores.
```
Fixed

AnnsAnns commented on cpu/riscv_common/irq_arch.c:
```diff
@@ -96,28 +96,47 @@ __attribute((used)) static void handle_trap(uword_t mcause)
     uword_t trap = mcause & CPU_CSR_MCAUSE_CAUSE_MSK;
 
 #ifdef DEVELHELP
+#   define
```
Oh there used to be a define there I used for something :sweat_smile:

AnnsAnns commented on boards/rpi-pico-2-arm/dist/openocd.cfg:
```diff
@@ -0,0 +1,10 @@
+echo "Make sure to use the Raspberry Pi OpenOCD version!"
+echo "For further details see:"
+echo "https://datasheets.raspberrypi.com/pico/getting-started-with-pico.pdf#%5B%7B%22num%22%3A18%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C115%2C604.93%2Cnull%5D"
```
fixed

AnnsAnns commented on boards/rpi-pico-2-riscv/include/periph_conf.h:
```diff
@@ -0,0 +1,62 @@
+ * @author          Tom Hert <git@annsann.eu>
+ */
+
+#include "kernel_defines.h"
+#include <stdint.h>
+#include "periph_cpu.h"
```
fixed

AnnsAnns commented on cpu/riscv_common/include/xh3irq.h:
```diff
@@ -0,0 +1,88 @@
+
+#pragma once
+
+#include <stdio.h>
+#include "cpu_conf.h"
+#include "panic.h"
```
fixed

AnnsAnns commented on cpu/riscv_common/irq_arch.c:
```diff
@@ -83,20 +87,64 @@ void riscv_irq_init(void)
 
+#ifdef DEVELHELP
+    printf("Trap: mcause=0x%" PRIx32 " mepc=0x%lx mtval=0x%lx\n",
+        (uint32_t)mcause, read_csr(mepc), read_csr(mtval));
```
fixed

AnnsAnns commented on cpu/riscv_common/irq_arch.c:
```diff
@@ -83,20 +87,64 @@ void riscv_irq_init(void)
+        };
+
+        uword_t cause_code = mcause & ~MCAUSE_INT;
+        if (cause_code <= (sizeof(error_messages) / sizeof(error_messages[0]) - 1)) {
```
fixed

crasbe commented:
I resolved some of the review comments that were fixed, but here are the ones that either haven't been addressed yet or where I'm not sure if they should be marked as resolved yet.

https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2391828261
https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2391859578
https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2391881875
https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2392372415
https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2392381003
https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2394888094

AnnsAnns commented on cpu/rp2350_common/periph/uart.c:
```diff
@@ -0,0 +1,237 @@
+
+#ifdef RP2350_USE_RISCV
+#  include "xh3irq.h"
+#endif
```
I moved the arch specific parts of the uart into a special static inline compat layer now (In anticipation that one day in the far future this could might also get compatible with the pico 1)

AnnsAnns commented on cpu/riscv_common/periph/xh3irq.c:
```diff
@@ -0,0 +1,81 @@
+     * 512-bit vector. The window is indexed by the 5 LSBs of the write data.
+     */
+    __asm__ volatile(
+        "csrs 0xbe0, %0\n"
+        : : "r"(index | (mask << INTERRUPT_ARRAY_MASK_OFFSET))
+
+    );
```
fixed

AnnsAnns commented on cpu/riscv_common/irq_arch.c:
```diff
@@ -83,20 +87,45 @@ void riscv_irq_init(void)
+        };
+
+        uword_t cause_code = mcause & ~MCAUSE_INT;
+        if (cause_code <= 0xb && error_messages[cause_code] != NULL) {
```
fixed

crasbe commented:
The reason for the failed build is that the `RP2350.h` also defines `BIT0`, but in a struct:
<img width="1289" height="874" alt="image" src="https://github.com/user-attachments/assets/292fd93c-40d9-4d24-ad46-df4b083015e6" />

https://github.com/raspberrypi/pico-sdk/blob/a1438dff1d38bd9c65dbd693f0e5db4b9ae91779/src/rp2_common/cmsis/stub/CMSIS/Device/RP2350/Include/RP2350.h#L674-L695

I guess there are two ways around that: either do a lot of patches in RIOT or do a single patch of the RP2350.h for a peripheral that is not used (yet) 🤔😅

AnnsAnns commented:
> > > The reason for the failed build is that the `RP2350.h` also defines `BIT0`, but in a struct: <img alt="image" width="1289" height="874" src="https://private-user-images.githubusercontent.com/1836007/496628592-292fd93c-40d9-4d24-ad46-df4b083015e6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk0MTAyMDEsIm5iZiI6MTc1OTQwOTkwMSwicGF0aCI6Ii8xODM2MDA3LzQ5NjYyODU5Mi0yOTJmZDkzYy00MGQ5LTRkMjQtYWQ0Ni1kZjRiMDgzMDE1ZTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTAwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEwMDJUMTI1ODIxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjZiZTU5OTIwMTU0ZGY2ZDc5YmQ2MTlhODg4Mzk3MThhZmM5MTYxMzNhMTFiM2JiZTIyYTNkMGYxYTFjOTA4ZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.VGEkgcmiWxzT0zSkqRm1DW4-g4nwBV_ZoJ95gx_NmbE">
> > > https://github.com/raspberrypi/pico-sdk/blob/a1438dff1d38bd9c65dbd693f0e5db4b9ae91779/src/rp2_common/cmsis/stub/CMSIS/Device/RP2350/Include/RP2350.h#L674-L695
> > > I guess there are two ways around that: either do a lot of patches in RIOT or do a single patch of the RP2350.h for a peripheral that is not used (yet) 🤔😅
> > 
> > 
> > bleeeeh 😵‍💫 Presumably because of using bit.h now?
> 
> I guess. But it's good that it happens now, because the RP2350 "BIT" stuff is not at all the same as the `bit.h` stuff. That might have caused weird errors.
> 
> Also you can use RIOT's patch feature to patch the file, similar to what we did with the CMakeLists of the picotool.

Im a bit confused, doing a clean build I dont get this error?

AnnsAnns committed:
pkg/picosdk: Add patch for bit overlap

AnnsAnns commented:
Something broke in the fixups :disappointed:

AnnsAnns commented:
It appears like the culprit was similar to the issue I had before, the riscv interrupt entry expects the scheduler to be active before going into interrupts, since we initialized the rp2350 uart, etc. before that it could cause the uart to trigger and already queue before exiting the cpu_init function. Simply enabling interrupts as the last part of the cpu_init appears to mitigate that issue.

AnnsAnns commented on cpu/rp2350_common/periph/uart.c:
```diff
@@ -225,6 +225,10 @@ void isr_handler(uint8_t num)
 
     if (status & UART_UARTMIS_RXMIS_BITS) {
         uint32_t data = dev->UARTDR;
+        if (data != 0) {
+            printf("UART%u: Received data: 0x%02lx ('%c')\n", num, data,
+                   (data >= 32 && data <= 126) ? (char)data : '.');
+        }
```
Ah heck

AnnsAnns commented on cpu/rp2350_riscv/Makefile.include:
```diff
@@ -0,0 +1,18 @@
+CFLAGS += -D$(CPU_MODEL)
+CFLAGS += -DROM_START_ADDR=$(ROM_START_ADDR)
+CFLAGS += -DRAM_START_ADDR=$(RAM_START_ADDR)
+CFLAGS += -Wno-error
```
fixed 8317537bd6

AnnsAnns commented:
> I resolved some of the review comments that were fixed, but here are the ones that either haven't been addressed yet or where I'm not sure if they should be marked as resolved yet.
> 
> [#21753 (comment)](https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2391828261) [#21753 (comment)](https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2391859578) [#21753 (comment)](https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2391881875) [#21753 (comment)](https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2392372415) [#21753 (comment)](https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2392381003) [#21753 (comment)](https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2394888094)

Tracking:

- [x] [#21753 (comment)](https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2391828261)
- [ ] [#21753 (comment)](https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2391859578)
  - Not yet 
- [x] [#21753 (comment)](https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2391881875)
- [ ] [#21753 (comment)](https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2392372415)
  - Unsure how to handle currently
- [x] [#21753 (comment)](https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2392381003)
- [x] [#21753 (comment)](https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2394888094)

AnnsAnns commented on cpu/rp2350_common/periph/uart.c:
```diff
@@ -0,0 +1,243 @@
+
+    uart_init_pins(uart);
+
+    if (uart_mode(uart, UART_DATA_BITS_8, UART_PARITY_NONE, UART_STOP_BITS_1) != UART_OK) {
```
Ok so I must admit that I don't fully understand what you mean, what function do I define in the uart config? Maybe I'm missing something :sweat_smile:

crasbe replied:
I assumed that the UART configuration includes flags for the data bits, parity and stop bits but apparently it does not 🤔 

So my review comment is resolved.

AnnsAnns replied:
Hooray but I agree that its rather a bit weird that it doesnt

AnnsAnns commented on boards/rpi-pico-2-riscv/include/periph_conf.h:
Is a rpi-pico-2-common folder something that makes sense in this context currently?

crasbe replied:
For boards, the common parts are in the subfolder `boards/common`, so in this case `boards/common/rpi-pico-2`.

AnnsAnns replied:
will do

crasbe commented on cpu/rp2350_common/include/compat_layer.h:
No?

AnnsAnns replied:
Yes 2f1dc02bb1ab7f1fda501352cc1825c8f082d916

I think Github is just bugging out?

crasbe commented on cpu/rp2350_common/include/compat_layer.h:
```diff
@@ -0,0 +1,57 @@
+ * @brief           Compatibility layer for different CPU architectures
+ *
+ * @author          Tom Hert <git@annsann.eu>
+ */
+
+
+#include "periph_cpu.h"
```
```suggestion
 */

#include "periph_cpu.h"
```

AnnsAnns commented:
Its hard to keep track with Github but I think I should have covered all feedback, thank you for all the reviews so far :pray:

crasbe commented on boards/rpi-pico-2-arm/Makefile:
```diff
@@ -1,3 +1,6 @@
 MODULE = board
 
+# Use the common board.
+DIRS += $(RIOTBOARD)/common/rpi-pico-2
+
```
This can be avoided by using the "new mechanism" I implemented (and have not documented I guess 👀 ) for using common boards. You can use this commit as reference for now: https://github.com/RIOT-OS/RIOT/pull/21332/commits/b90316166359107f916657076eefba0f595dbe0e

Essentially you explicitly call the common `boards/common/bla/Makefile{,.dep,.include,.features}` in the `boards/Makefile{,.dep,.include,.features}`, so that you only have to specify the `USEMODULE += boards_common_bla` in your Makefile.

AnnsAnns commented on boards/Makefile:
```diff
@@ -9,5 +9,8 @@ endif
 ifneq (,$(filter boards_common_seeedstudio-xiao-nrf52840,$(USEMODULE)))
   DIRS += $(RIOTBOARD)/common/seeedstudio-xiao-nrf52840
 endif
+ifneq (,$(filter boards_common_rpi_pico_2,$(USEMODULE)))
+  DIRS += $(RIOTBOARD)/common/rpi-pico-2
+endif
```
I was only one letter off, I feel like thas fine, its basically a 25/26 success rate here, give me some slack :pensive: (I didnt notice that comment)

crasbe reviewed (commented):
I think though you can do an intermediate squash already. Or you can squash everything at the very end, however you prefer.

Hopefully I'll get around testing it on real hardware soon.

AnnsAnns commented:
Will do a squash tomorrow, I also updated the UART to be fairly feature complete now after a full day of debugging a really weird UART reading issue only to find out that the pins were swapped on my adapter (but for some reason UART writing still worked, its really cursed and I dont want to question it anymore)

AnnsAnns committed:
cpu/rp2350_riscv: add xh3irq support

AnnsAnns committed:
cpu/rp2350_riscv: move to shared rp2350_common

AnnsAnns committed:
cpu/rp2350_common: change register calculation functions to pointers

AnnsAnns commented:
Not really sure what triggers this but rebase kept creating multiple identical commits in the process of rebasing.

After lunch I'll diff the pushed commits just to ensure that I didn't accidentally regress anything in the process.

crasbe commented on cpu/rp2350_common/periph/uart.c:
```diff
@@ -0,0 +1,275 @@
+        calculate_gpio_pad_register_addr(uart_config[uart].rx_pin),
+        PADS_BANK0_GPIO0_IE_BITS);
+
+    dev->UARTIBRD = IBRD;
+    dev->UARTFBRD = FBRD;
```
Why is the baud rate set when initializing the GPIOs?

crasbe commented on cpu/rp2350_common/periph/uart.c:
```diff
@@ -0,0 +1,275 @@
+
+    /* enable RX and IRQs, if needed */
+    if (rx_cb != NULL) {
+        _irq_enable(uart);
```
It usually makes sense to enable the interrupts *after* setting the registers and clearing possibly pending interrupts.

crasbe commented on cpu/rp2350_common/periph/uart.c:
```diff
@@ -0,0 +1,275 @@
+    }
+}
+
+void uart_poweroff(uart_t uart)
```
Are the interrupts disabled in this function?

crasbe commented on cpu/rp2350_common/vectors.c:
```diff
@@ -14,13 +14,27 @@
+#endif
+
+#ifndef WEAK_DEFAULT
+#   define WEAK_DEFAULT __attribute__((weak, alias("dummy_handler")))
```
```suggestion
#  define WEAK_DEFAULT __attribute__((weak, alias("dummy_handler")))
```

Whitespace error 😈

crasbe commented on cpu/rp2350_riscv/cpu.c:
```diff
@@ -0,0 +1,37 @@
+ * @author      Tom Hert <git@annsann.eu>
+ * @}
+ */
+
+#include "board.h"
+#include "cpu.h"
+#include "clock_conf.h"
+#include "kernel_init.h"
+#include "periph/init.h"
+#include "periph/uart.h"
+#include "periph_conf.h"
+
+#include <sys/unistd.h>
```
```suggestion

#include <sys/unistd.h>

#include "board.h"
#include "cpu.h"
#include "clock_conf.h"
#include "kernel_init.h"
#include "periph/init.h"
#include "periph/uart.h"
#include "periph_conf.h"
```

crasbe commented on cpu/rp2350_riscv/include/cpu_conf.h:
```diff
@@ -0,0 +1,34 @@
+ *
+ * @author          Tom Hert <git@annsann.eu>
+ */
+
+#include "cpu_conf_common.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define CPU_DEFAULT_IRQ_PRIO 1u
+#define CPU_IRQ_NUMOF 52u
+#define NUM_PMP_ENTRIES 16u
```
Undocumented boi.

AnnsAnns committed:
cpu/rp2350_arm: multicore support

AnnsAnns committed:
boards/rpi-pico-2-*: create common board folder

AnnsAnns commented:
Also since multicore can't/isn't being tested right now, here is some proof of it running: (ARM here, RISC-V was also tested)
<img width="760" height="284" alt="image" src="https://github.com/user-attachments/assets/6f918983-9cc0-41a6-ab60-1721fb46eb0f" />

AnnsAnns committed:
cpu/rp2350/gpio: Add PIO support

AnnsAnns committed:
cpu/rp2350: Add PIO assembly macros

AnnsAnns commented:
> This PR grew a lot. Do you think it would be possible to split it into smaller PRs? Obviously this only makes sense if that is less work for you than for the reviewer to go through the current PR.

It depends, afaik, the review was till 3b96eb3baca6dc9e34537d0b1ac4654463f22787 - I could remove the multicore/PIO stuff from this PR again to make it a bit easier though

AnnsAnns committed:
cpu/rp2350_common: reduce stack size to main stack default

AnnsAnns commented:
So this PR has gotten a bit large and I am not totally sure which points need to still be addressed, https://github.com/RIOT-OS/RIOT/pull/21753#issuecomment-3360261811 seems to have been addressed mostly

crasbe commented:
Considering the RP2350 support is still experimental, I'd propose to YOLO it and see if anything transpires afterwards.

crasbe commented on cpu/rp2350_common/core.c:
```diff
@@ -0,0 +1,134 @@
+    core_1_stack[1] = (uint32_t) arg;
+
+    uint32_t seq = 0;
+    /** We iterate through the cmd_sequence till we covered every param
```
```suggestion
    /* We iterate through the cmd_sequence till we covered every param
```

crasbe commented on cpu/rp2350_common/core.c:
```diff
@@ -0,0 +1,134 @@
+        (uint32_t) _core1_trampoline,
+    };
+
+    /**
```
```suggestion
    /*
```

The `/**` is used to let Doxygen know that a comment block starts. Doxygen does not evaluate these comments, so it's not necessary.

crasbe commented on cpu/rp2350_common/core.c:
```diff
@@ -0,0 +1,134 @@
+        uint32_t cmd = cmd_sequence[seq];
+        /* If the cmd is 0 we need to drain the READ FIFO first*/
+        if (cmd == 0) {
+            /**
```
```suggestion
            /*
```

crasbe commented on cpu/rp2350_common/periph/uart.c:
```diff
@@ -0,0 +1,270 @@
+        calculate_gpio_pad_register_addr(uart_config[uart].rx_pin),
+        PADS_BANK0_ISO_BITS);
+
+    /* Set Input Enable Flag  */
```
```suggestion
    /* Set Input Enable Flag */
```

crasbe commented on cpu/riscv_common/irq_arch.c:
```diff
@@ -80,20 +84,63 @@ void riscv_irq_init(void)
+                printf("Machine Cause indicates that 0x%lx"
+                       " is a reserved or custom cause code\n",
+                       cause_code
+                );
```
```suggestion
                        cause_code, error_messages[cause_code]);
            } else {
                printf("Machine Cause indicates that 0x%lx"
                       " is a reserved or custom cause code\n",
                       cause_code);
```

Similar to the proposel of Uncrustify.

crasbe commented on boards/rpi-pico-2-riscv/Makefile.include:
```diff
@@ -0,0 +1,7 @@
+# JLink isnt tested yet on RP2350
+# ifeq ($(PROGRAMMER),jlink)
+#   JLINK_DEVICE = RP2350_M33_0
+# endif
```
```suggestion
# The J-Link Software has to be at least V8.10c from 16.10.2024
ifeq ($(PROGRAMMER),jlink)
  JLINK_DEVICE = RP2350_RV32_0
endif
```

crasbe reviewed (commented):
It seems like the J-Link support for the RP2350 is not great currently.

Also, the same issue as described here occurs when flashing the RISC-V core:
https://forum.segger.com/thread/9977-rp2350-extremely-poor-performance-after-flashing/?postID=35143&highlight=RP2350#post35143

See the video. At some point the flashing gets faster lol. This is `examples/basic/blinky`
https://github.com/user-attachments/assets/3778fb57-a01f-4558-a1c9-bfaef9123bdb

By the way: is blinky supposed to blink that fast? No timer support as it seems.

crasbe commented on cpu/rp2350_common/core.c:
```diff
@@ -0,0 +1,134 @@
+             * gets discarded (called multicore_fifo_drain in chapter 5.3)
+             */
+            while (SIO->FIFO_ST & 1<<SIO_FIFO_READ_VALID_BIT) {
+                (void) SIO->FIFO_RD; /* Table 39 FIFO_RD*/
```
You missed this one :P

crasbe commented on cpu/rp2350_common/periph/uart.c:
```diff
@@ -0,0 +1,270 @@
+    SIO->GPIO_OE_CLR = 1LU << uart_config[uart].tx_pin;
+    if (ctx[uart].rx_cb) {
+        /* gpio_reset_all_config(uart_config[uart].rx_pin); */
+    }
```
Perhaps it would be good to add some documentation about what the TODO is about then 🤔

AnnsAnns commented:
Rebased everything and integrated the review changes into the commits they relate to 👍 If needed, I also backed it up beforehand: https://github.com/AnnsAnns/RIOT/tree/double_backup_two

crasbe reviewed (commented):
These points are still open:

- [x] https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2413897662
- [x] https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2840793485
- [x] https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2843254011
- [x] https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2840850972

crasbe commented:
> Rebased everything and integrated the review changes into the commits they relate to 👍 If needed, I also backed it up beforehand: https://github.com/AnnsAnns/RIOT/tree/double_backup_two

It seems like something broke for the RISC-V core:
```
cbuec@W11nMate:~/RIOTstuff/riot-vanilla/RIOT$ BOARD=rpi-pico-2-riscv make -C examples/basic/blinky/
make: Entering directory '/home/cbuec/RIOTstuff/riot-vanilla/RIOT/examples/basic/blinky'
Building application "blinky" for "rpi-pico-2-riscv" with CPU "rp2350_riscv".

"make" -C /home/cbuec/RIOTstuff/riot-vanilla/RIOT/pkg/mpaland-printf/
"make" -C /home/cbuec/RIOTstuff/riot-vanilla/RIOT/build/pkg/mpaland-printf -f /home/cbuec/RIOTstuff/riot-vanilla/RIOT/pkg/mpaland-printf/mpaland-printf.mk
"make" -C /home/cbuec/RIOTstuff/riot-vanilla/RIOT/pkg/picosdk/
"make" -C /home/cbuec/RIOTstuff/riot-vanilla/RIOT/boards
"make" -C /home/cbuec/RIOTstuff/riot-vanilla/RIOT/boards/common/init
"make" -C /home/cbuec/RIOTstuff/riot-vanilla/RIOT/boards/common/rpi-pico-2
"make" -C /home/cbuec/RIOTstuff/riot-vanilla/RIOT/boards/rpi-pico-2-riscv
"make" -C /home/cbuec/RIOTstuff/riot-vanilla/RIOT/core
"make" -C /home/cbuec/RIOTstuff/riot-vanilla/RIOT/core/lib
"make" -C /home/cbuec/RIOTstuff/riot-vanilla/RIOT/cpu/rp2350_riscv
"make" -C /home/cbuec/RIOTstuff/riot-vanilla/RIOT/cpu/riscv_common
"make" -C /home/cbuec/RIOTstuff/riot-vanilla/RIOT/cpu/riscv_common/periph
"make" -C /home/cbuec/RIOTstuff/riot-vanilla/RIOT/cpu/rp2350_common
/home/cbuec/RIOTstuff/riot-vanilla/RIOT/cpu/rp2350_common/core.c: In function 'core1_init':
/home/cbuec/RIOTstuff/riot-vanilla/RIOT/cpu/rp2350_common/core.c:57:21: error: '_isr_vectors' undeclared (first use in this function)
   57 |         (uint32_t) &_isr_vectors,
      |                     ^~~~~~~~~~~~
/home/cbuec/RIOTstuff/riot-vanilla/RIOT/cpu/rp2350_common/core.c:57:21: note: each undeclared identifier is reported only once for each function it appears in
make[3]: *** [/home/cbuec/RIOTstuff/riot-vanilla/RIOT/Makefile.base:157: /home/cbuec/RIOTstuff/riot-vanilla/RIOT/examples/basic/blinky/bin/rpi-pico-2-riscv/rp2350_common/core.o] Error 1
make[2]: *** [/home/cbuec/RIOTstuff/riot-vanilla/RIOT/Makefile.base:31: ALL--/home/cbuec/RIOTstuff/riot-vanilla/RIOT/cpu/rp2350_common] Error 2
make[1]: *** [/home/cbuec/RIOTstuff/riot-vanilla/RIOT/Makefile.base:31: ALL--/home/cbuec/RIOTstuff/riot-vanilla/RIOT/cpu/rp2350_riscv] Error 2
make: *** [/home/cbuec/RIOTstuff/riot-vanilla/RIOT/examples/basic/blinky/../../../Makefile.include:743: application_blinky.module] Error 2
make: Leaving directory '/home/cbuec/RIOTstuff/riot-vanilla/RIOT/examples/basic/blinky'
```

crasbe commented on cpu/riscv_common/irq_arch.c:
```diff
@@ -105,6 +150,9 @@ static void handle_trap(uword_t mcause)
             }
+            if (IS_ACTIVE(MODULE_PERIPH_XH3IRQ)) {
+                xh3irq_handler();
+            }
```
Too slow: https://github.com/RIOT-OS/RIOT/pull/21753#discussion_r2845853932 :P

crasbe replied:
![giphy](https://github.com/user-attachments/assets/8afe7efd-d530-4d42-84d7-2a4f7e0eec73)

crasbe reviewed (commented):
Merging the proposals of Teufelchen and me wasn't quite necessary 😅

AnnsAnns committed:
cpu/rp2350: provide pmp

cpu/rp2350: increase stack size

fix: pmp is actually just 8, the other 8 are a lie

fix: rp2350-e6 breaks PMP completely

crasbe commented on cpu/rp2350_common/core.c:
```diff
@@ -0,0 +1,130 @@
+        1,
+        /*
+        * Get the vector table pointer, this is needed to set the VTOR register on core1
+        */
```
```suggestion
        /* Get the vector table pointer, this is needed to set the VTOR register on core1 */
```

crasbe commented on cpu/rp2350_common/core.c:
```diff
@@ -0,0 +1,130 @@
+        * We allocate a stack "locally" instead of in the linker script
+        * since that would require changes to the base cortexm script
+        * which sound complicated to do on a per-cpu basis
+        */
```
```suggestion
         * We allocate a stack "locally" instead of in the linker script
         * since that would require changes to the base cortexm script
         * which sound complicated to do on a per-cpu basis
         */
```

crasbe commented on cpu/rp2350_arm/Makefile.features:
```diff
@@ -1,9 +1,7 @@
 CPU_CORE    := cortex-m33
 CPU_FAM     := RP2350
 CPU_ARCH    = armv8m
```
```suggestion
CPU_CORE := cortex-m33
CPU_FAM  := RP2350
CPU_ARCH := armv8m
```

AnnsAnns committed:
boards/rpi-pico-2: rename arm pico 2 version to rpi-pico-2-arm

AnnsAnns committed:
cpu/rp2350_riscv: multicore support

mguetschow commented on boards/common/rpi-pico-2/include/periph_conf.h:
```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+/**
+ * @ingroup         boards_rpi_pico_2_arm
```
```suggestion
 * @ingroup         boards_rpi_pico_2
```

mguetschow reviewed (commented):
Weak approve since I have only skimmed over it right now.

Thanks for doing this!
